### PR TITLE
[Yosegi-10] Balance compression efficiency and CPU resources in write…

### PR DESCRIPTION
**Pull request by koijima**
### What is the necessity of this update? What is updated?
In order to balance processing time and data size in write processing.
Added function to adjust compression level from compression ratio.

https://yosegi.atlassian.net/browse/YOSEGI-10

### How did you do the test? (Not required for updating documents)
Unit test added.

### CODE REVIEW

File: src/main/java/jp/co/yahoo/yosegi/binary/ColumnBinaryMakerConfig.java [hunk 1]
```diff
@@ -94,6 +94,7 @@ public ColumnBinaryMakerConfig() throws IOException {
     bytesMakerClass = FindColumnBinaryMaker.get( DumpBytesColumnBinaryMaker.class.getName() );
 
     compressionPolicy = CompressionPolicy.DEFAULT;
+    allowedRatio = 1.15d;
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/ColumnBinaryMakerConfig.java]: I think that it is better not to specify this value directly but to be able to set it by the external setting file.
However, if this parameter realizes good performance, it seems useful to set an initial setting.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/DumpArrayColumnBinaryMaker.java [hunk 2]
```diff
@@ -73,8 +76,13 @@ public ColumnBinary toBinary(
       }
     }
 
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/DumpArrayColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/DumpBooleanColumnBinaryMaker.java [hunk 3]
```diff
@@ -76,7 +79,13 @@ public ColumnBinary toBinary(
       }
     }
 
-    byte[] compressData = currentConfig.compressorClass.compress( binary , 0 , binary.length );
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/DumpBooleanColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/DumpBytesColumnBinaryMaker.java [hunk 4]
```diff
@@ -75,7 +78,13 @@ public ColumnBinary toBinary(
       columnList.add( objList.size() - 1 );
     }
     byte[] binaryRaw = convertBinary( columnList , objList , currentConfig , totalLength );
-    byte[] binary = currentConfig.compressorClass.compress( binaryRaw , 0 , binaryRaw.length );
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/DumpBytesColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/DumpUnionColumnBinaryMaker.java [hunk 5]
```diff
@@ -151,8 +161,13 @@ public ColumnBinary toBinary(
       wrapBuffer.put( ColumnTypeFactory.getColumnTypeByte( column.get(i).getType() ) );
     }
 
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/DumpUnionColumnBinaryMaker.java]: What is "c0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeDoubleColumnBinaryMaker.java [hunk 6]
```diff
@@ -346,8 +348,13 @@ public ColumnBinary toBinary(
     indexMaker.create( indexArray , binaryRaw , 0 , indexLength , order );
     dicMaker.create( dicList , binaryRaw , indexLength , dicLength , order );
 
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeDoubleColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeDumpLongColumnBinaryMaker.java [hunk 7]
```diff
@@ -718,8 +720,13 @@ public ColumnBinary toBinary(
     binaryMaker.create(
         valueArray , binaryRaw , nullBinaryLength , valueLength , order , rowCount );
 
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeDumpLongColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeDumpStringColumnBinaryMaker.java [hunk 8]
```diff
@@ -458,8 +460,13 @@ public ColumnBinary toBinary(
     for ( int i = 0 ; i < objList.length ; i++ ) {
       wrapBuffer.put( objList[i] );
     }
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeDumpStringColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeFloatColumnBinaryMaker.java [hunk 9]
```diff
@@ -340,8 +342,13 @@ public ColumnBinary toBinary(
     indexMaker.create( indexArray , binaryRaw , 0 , indexLength , order );
     dicMaker.create( dicList , binaryRaw , indexLength , dicLength , order );
 
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeFloatColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeLongColumnBinaryMaker.java [hunk 10]
```diff
@@ -725,8 +727,13 @@ public ColumnBinary toBinary(
     indexMaker.create( indexArray , binaryRaw , 0 , indexLength , order );
     dicMaker.create( dicList , binaryRaw , indexLength , dicLength , order );
 
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeLongColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeStringColumnBinaryMaker.java [hunk 11]
```diff
@@ -621,8 +623,13 @@ public ColumnBinary toBinary(
     for ( byte[] obj : stringList ) {
       wrapBuffer.put( obj );
     }
-    byte[] compressBinaryRaw =
-        currentConfig.compressorClass.compress( binaryRaw , 0 , binaryRaw.length , DataType.TEXT );
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeOptimizeStringColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeRangeDumpDoubleColumnBinaryMaker.java [hunk 12]
```diff
@@ -106,18 +108,23 @@ public ColumnBinary toBinary(
     byte byteOrderByte = order == ByteOrder.BIG_ENDIAN ? (byte)0 : (byte)1;
     parentsBinaryRaw[column.size()] = byteOrderByte;
 
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeRangeDumpDoubleColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeRangeDumpFloatColumnBinaryMaker.java [hunk 13]
```diff
@@ -103,18 +105,23 @@ public ColumnBinary toBinary(
     byte byteOrderByte = order == ByteOrder.BIG_ENDIAN ? (byte)0 : (byte)1;
     parentsBinaryRaw[column.size()] = byteOrderByte;
 
+    CompressResult compressResult = compressResultNode.getCompressResult(
+        this.getClass().getName() ,
+        "c0"  ,
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/binary/maker/UnsafeRangeDumpFloatColumnBinaryMaker.java]: What is "C0"?
This constant value should be named to easy to understand.

File: src/main/java/jp/co/yahoo/yosegi/block/PushdownSupportedBlockWriter.java [hunk 14]
```diff
@@ -85,6 +87,12 @@ public void setup( final int blockSize , final Configuration config ) throws IOE
       defaultConfig.compressorClass =
           FindCompressor.get( config.get( "spread.column.maker.default.compress.class" ) );
     }
+    if ( config.containsKey( "compress.optimize.allowed.ratio" ) ) {
+      double allowedRatio = config.getDouble( "compress.optimize.allowed.ratio" , 1.25d );
```
Comment (yohto) [inline src/main/java/jp/co/yahoo/yosegi/block/PushdownSupportedBlockWriter.java]: I think that it is better not to specify this value directly but to be able to set it by the external setting file.
But, if this parameter realizes good performance, it seems good to set initial setting.

**Comment by koijima** [comment]
> I found the value named “c0”.
> I think this constant value should be named or add some comment for easy to understand.

This is not a constant.
Multiple compression processes are done in the class.
c0 is a unique name representing compression processing in that class.

**Comment by yohto** [comment]
I understand.

**Comment by yohto** [comment]
LGTM

**Comment by koijima** [comment]
> I think that it is better not to specify this value directly but to be able to set it by the external setting file.
> However, I can understand it is for initial setting and it seems to realize good performance might be.

I think so, too.
It would be nice to have a class to hold constant classes.
However, because it is different from the contents of this time, I made a challenge to JIRA.
https://yosegi.atlassian.net/browse/YOSEGI-13

**Pull request merged**